presto-docs/src/main/sphinx/admin/cte-materialization.rst

```diff
@@ -0,0 +1,103 @@
+
+How to use CTE Materialization
+--------------
+
+The following configurations and session properties are provided to enable CTE materialization and fine-tune its settings.
+
+``cte-materialization-strategy``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``string``
+    * **Allowed values:** ``ALL``, ``NONE``, ``HEURISTIC`, ``HEURISTIC_COMPLEX_QUERIES_ONLY``
+    * **Default value:** ``NONE``
+
+    Specifies the strategy to use for materializing Common Table Expressions (CTEs) in queries.
+    ``NONE`` indicates that no CTEs will be materialized.
+    ``ALL`` indicates that all CTEs in the query will be materialized.
+    ``HEURISTIC` greedily materializes the earliest parent CTE which is repeated >= cte_heuristic_replication_threshold times
+    ``HEURISTIC_COMPLEX_QUERIES_ONLY` greedily materializes the earliest parent CTE which meets the heuristic criteria and has a join or aggregate
+    This can also be specified on a per-query basis using the ``cte_materialization_strategy`` session property.
+
+``cte-heuristic-replication-threshold``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``integer``
+    * **Minimum value:** ``0``
+    * **Default value:** ``4``
+
+    Used with CTE Materialization Strategy is set to ``HEURISTIC` or ``HEURISTIC_COMPLEX_QUERIES_ONLY``. CTES are only materialized if they are used greater than or equal to this number
+    This can also be specified on a per-query basis using the ``cte_heuristic_replication_threshold`` session property.
+
+``query.cte-hash-partition-count``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``integer``
+    * **Default value:** ``100``
+
+    The number of partitions to be used for materializing Common Table Expressions (CTEs) in queries.
+    This setting determines how many buckets or writers should be used when materializing the CTEs, potentially affecting the performance of queries involving CTE materialization.
+    A higher number of partitions might improve parallelism but also increases overhead in terms of memory and network communication.
+    Recommended value: 4 - 10x times the size of the cluster.
+    This can also be specified on a per-query basis using the ``cte_hash_partition_count`` session property.
+
+``query.cte-partitioning-provider-catalog``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``string``
+    * **Default value:** ``system``
+
+    The name of the catalog to be used and which provides custom partitioning for Common Table Expression (CTE) materialization.
+    This setting specifies which catalog should be used for CTE materialization and for determining how to partition the materialization of CTEs in queries.
+    This can also be specified on a per-query basis using the ``cte_partitioning_provider_catalog`` session property.
+
+``cte-filter-and-projection-pushdown-enabled``
```

Comment:
Would be great to have an example of this as well. Do you have something in a design doc ?

Reply:
https://github.com/prestodb/presto/issues/21638 There is a good example on this ticket